fml/message_loop_impl.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/message_loop_impl.h"
+#include "lib/ftl/build_config.h"
+
+#if OS_MACOSX
+
+#include "flutter/fml/platform/darwin/message_loop_darwin.h"
+using PlatformMessageLoopImpl = fml::MessageLoopDarwin;
+
+#else
+
+#error This platform does not have a message loop implementation.
+
+#endif
+
+namespace fml {
+
+ftl::RefPtr<MessageLoopImpl> MessageLoopImpl::Create() {
+  return ftl::MakeRefCounted<::PlatformMessageLoopImpl>();
+}
+
+MessageLoopImpl::MessageLoopImpl() : order_(0) {}
+
+MessageLoopImpl::~MessageLoopImpl() = default;
+
+void MessageLoopImpl::PostTask(ftl::Closure task, ftl::TimePoint target_time) {
```

Comment:
The destruction semantics aren't quite right.  Here are the expected semantics:

1. If the message loop has already been destroyed, PostTask should destruct |task| synchronously within this function.
2. If the message loop is in the process of shutting down, |task| should be destructed on the message loop's thread.
3. If the message loop is still running, |task| should be destructed after it is invoked.

The normal approach for implementing this functionality is to put the incoming task queue in its own thread-safe refcounted object.  When the message loop wants to shut down, it takes the incoming task queue lock, and sets a terminated flag in the incoming task queue and takes any incoming tasks.  From this point onward, any tasks that are posted to the incoming queue are immediately destructed because the terminated bit is set.

Meanwhile, after releasing the incoming task log, the message loop destructs all the pending tasks that it has that never got a chance to execute.